Removes hardcoded module loading order

yurydelendik opened the pull request:
In non-production environment, RequireJS is used to load all source files/modules. <del>For now I placed it in the external/requirejs/ folder (I did not decide yet if we want to pull it as devDependency)</del>

In production environment, pdf.js and pdf.worker.js bundles are created based on the order provided by the external/umdutils/verifier.js readDependencies.

yurydelendik committed:
Use RequireJS in the viewer, examples and tests.

yurydelendik committed:
Adds RequireJS to worker.

timvandermeij commented:
/botio-linux preview

yurydelendik committed:
Use UMD headers to detect module loading order.

timvandermeij commented:
I think getting RequireJS with `npm` and making it a development depency is a good idea. It will be easier to apply updates for the library and requires less code in the repository (ShellJS was also a good example of this).

yurydelendik commented on package.json:
```diff
@@ -4,10 +4,11 @@
     "rimraf": "^2.4.1",
     "shelljs": "~0.4.0",
     "typogr": "~0.6.5",
+    "wintersmith": "^2.0.0",
```
moved automatically by `npm install requirejs --save-dev` command

timvandermeij commented on examples/svgviewer/viewer.js:
```diff
@@ -53,4 +45,15 @@ PDFJS.getDocument(url).then(function(pdf) {
   }
+}
+
+
```
Nit: unnecessary newline

timvandermeij commented:
@yurydelendik Could you rebase this patch?

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
/botio test

yurydelendik commented:
/botio test

yurydelendik commented:
/botio test

timvandermeij commented:
Chrome fails after `eq test usmanm-bad-auto-fetch`.

yurydelendik commented:
> Chrome fails after eq test usmanm-bad-auto-fetch.

Looks like Chrome does not like this amount of importScripts (within requirejs timeout?). Replacing it with Promise.then callback.

/botio test

yurydelendik commented on src/worker_loader.js:
```diff
@@ -12,70 +12,20 @@
+  // If possible, using Promises queue instead of RequireJS setTimeout.
+  require.s.contexts._.nextTick = (function (promise) {
+    return promise.then.bind(promise);
+  })(Promise.resolve(undefined));
```
@Rob--W not sure if it will be an interest to you, but on Chrome, if I remove this hack, testing will start failing: requirejs somewhat stops resolving and worker code is not loaded. I rigged requirejs code with tracing and that shows that it stops issuing importScripts and/or execution of the requirejs logic stops.

Rob--W replied:
... after hours of debugging, I discovered the cause of this bug (the worker is being garbage collected), and reported it at https://code.google.com/p/chromium/issues/detail?id=572225.

Your work-around does not solve the root cause of the bug. To fix this bug, you have two options:

1. (easy, but quite hacky-ish): Override `importScripts` with synchronous XMLHttpRequest + `eval` (note: sync xhr is totally fine for worker threads) (note2: maybe add a fallback to `importScripts` when sync xhr fails for some (security?) reason)
2. Change the APIs / tests such that a reference to the `Worker` instance exists until it is no longer needed. This is hard, I naively tried to use `this._webWorker = worker;` in `display/api.js` without luck. Using `window['foo' + Math.random()] = worker;` did work, which shows that this issue is really caused by GC.

yurydelendik replied:
I added the hack listed in p.2. Thanks for finding this out.

yurydelendik committed:
Hack to avoid intermidiate Chrome failures during tests.

Remove when https://code.google.com/p/chromium/issues/detail?id=572225 is fixed.

yurydelendik commented:
/botio test

yurydelendik commented:
/botio test

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Thank you, this looks really good!
